sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -896,6 +896,25 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("SPARK-24012 Union of map and other compatible columns") {
```

Comment:
cc @gatorsmile , what's the policy for end-to-end tests? Shall we add it in both the sql golden file and `SQLQuerySuite`?

Reply:
yes. please add them to SQLQueryTestSuite

Reply:
discussed with @gatorsmile , we should put end-to-end test in a single place, and currently we encourage people to put SQL related end-to-end test in the SQL golden files. That is to say, we should remove this test from `SQLQuerySuite`.

In the meanwhile, a bug fix should also have a unit test. For this case, we should add a test case in `TypeCoercionSuite`. @liutang123 if you are not familiar with that test suite, please let us know, we can merge your PR first and add UT in `TypeCoercionSuite` in a followup.

Reply:
@cloud-fan , Yes, I am not familiar with TypeCoercionSuite. In order to save time, in my opinion, this PR can be merged first. Thanks a lot.

Reply:
OK, please remove this test and it's ready to go.